### PR TITLE
[4.0] Adding a node object interface

### PR DESCRIPTION
**Pull request by Hackwar**
This adds an interface for node objects. For now this only adds the interface, but these are supposed to be applied to CategoriesNode, MenuItem and a class for tags in the future. And of course wherever else they would fit.

### CODE REVIEW

**Comment by rdeutz** [comment]
looks good me, maybe a getRoot function could make sense

**Comment by Hackwar** [comment]
Adding a getRoot method would mean that the objects need to know about the whole tree and not just their direct children and their parent. I'd like to keep it as simple as possible and each node by default would only need to know the bare minimum.

**Comment by rdeutz** [comment]
ok, then you have to use getParent() so often till you not get a parent. Is also a way to go.

File: libraries/src/Tree/ImmutableNodeInterface.php [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+	 * @return  NodeInterface[]  The children
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function &getChildren($recursive = false);
+
+	/**
+	 * Get the parent of this node
+	 *
+	 * @return  NodeInterface
```
Comment (mbabker) [inline libraries/src/Tree/ImmutableNodeInterface.php]: Should be nullable as it isn't guaranteed every node will have a parent (i.e. a root node)

**Comment by mbabker** [comment]
```php
public function getRoot(): object
{
    $root = $this->getParent();

    if (!$root) {
        return $this;
    }

    while ($root->getParent()) {
        $root = $root->getParent();
    }

    return $root;
}
```

**Comment by Hackwar** [comment]
So, is the consensus that you want such a getRoot method? I personally would want to keep it rather simple and leave such methods out which can be created with the other methods of the interface. But in any case, just say which way to go and I'll adapt it. It would just be nice to move forward with this.

**Comment by mbabker** [comment]
Personally I'd say add it, `getRoot()` is a rather simple thing and a similar thing does exist in most other tree oriented packages I work with.

**Comment by Hackwar** [comment]
getRoot has been added. Can we simply merge this or do we need "testers"?

**Pull request merged**